security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -335,6 +356,52 @@ func (sc *SecretCache) GenerateSecret(ctx context.Context, connectionID, resourc
 	return ns, nil
 }
 
+func (sc *SecretCache) addFileWatcher(file string, key ConnKey) {
+	// TODO(ramaraochavali): add integration test for file watcher functionality.
+	// Check if this file is being already watched, if so ignore it. FileWatcher has the functionality of
+	// checking for duplicates. This check is needed here to avoid processing duplicate events for the same file.
+	sc.certMutex.Lock()
+	_, exists := sc.fileCerts[file]
+	if !exists {
+		sc.fileCerts[file] = struct{}{}
+	}
+	sc.certMutex.Unlock()
+	// File is not being watched, start watching now and trigger key push.
+	if !exists {
+		cacheLog.Debugf("adding watcher for file %s", file)
+		if err := sc.certWatcher.Add(file); err != nil {
+			cacheLog.Errorf("error adding watcher for file %s, Skipping watches ", file)
+			return
+		}
+		go func() {
+			var timerC <-chan time.Time
+			for {
+				select {
+				case <-timerC:
+					timerC = nil
+					if val, ok := sc.secrets.Load(key); ok {
+						// Trigger the callback that pushes the secrets to proxy.
+						secret := val.(model.SecretItem)
+						cacheLog.Debugf("file changed %s, triggering push to proxy", file)
+						sc.callbackWithTimeout(key, &secret)
+					}
+				case e := <-sc.certWatcher.Events(file):
+					if e.Op&fsnotify.Remove == fsnotify.Remove {
+						// File is deleted. Stop watching
+						cacheLog.Debugf("file deleted %s, removing the watcher", file)
+						_ = sc.certWatcher.Remove(file)
+						return
+					}
+					// Use a timer to debounce watch updates
+					if timerC == nil {
+						timerC = time.After(100 * time.Millisecond) // TODO: Make this configurable if needed.
```

Comment:
So you introduced this to prevent two changing events of the private key and cert files to cause two secret push, correct? But there will still be two pushing events, with the same updated key and cert?

Reply:
Correct. There will be a push that sends both updated files because the push looks at the passed in secret item